qiskit/synthesis/linear/linear_circuits_utils.py

```diff
@@ -0,0 +1,103 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Utility functions for handling linear reversible circuits."""
+
+import copy
+from typing import Callable
+import numpy as np
+from qiskit import QuantumCircuit
+from qiskit.exceptions import QiskitError
+from qiskit.circuit.exceptions import CircuitError
+from . import calc_inverse_matrix, check_invertible_binary_matrix
+
+
+def transpose_cx_circ(qc: QuantumCircuit):
+    """Takes a circuit having only CX gates, and calculates its transpose.
+    This is done by recursively replacing CX(i, j) with CX(j, i) in all instructions.
+
+    Args:
+        qc: a QuantumCircuit containing only CX gates.
+
+    Returns:
+        QuantumCircuit: the transposed circuit.
+
+    Raises:
+        CircuitError: if qc has a non-CX gate.
+        CircuitError: if a CX gate is not on two qubits.
+    """
+    tranposed_circ = QuantumCircuit(qc.qubits, qc.clbits, name=qc.name + "_transpose")
+    for instruction in reversed(qc.data):
+        if instruction.operation.name != "cx":
+            raise CircuitError("The circuit contains non-CX gates.")
+        if instruction.operation.num_qubits != 2:
+            raise CircuitError("The circuit contains a CX gates which is not on two qubits.")
+        tranposed_circ._append(instruction.replace(qubits=reversed(instruction.qubits)))
+    return tranposed_circ
+
+
+def optimize_cx_4_options(function: Callable, mat: np.ndarray, optimize_count: bool = True):
```

Comment:
It would be good if could implement this function on top of the generic high-level-synthesis plugin-manager API. I am tempted to suggest to remove `optimize_cx_4_options` from this PR.

Reply:
Since all of the functions in this PR are utility functions, they are not added to the Qiskit documentation.
So we can add this function now, and actually use it in the high level synthesis transpiler pass in a following PR.

Reply:
This sounds reasonable, even if we mght need to massage this function a bit in the future.